pkg/sfu/receiver.go

```diff
@@ -360,52 +334,80 @@ func (w *WebRTCReceiver) writeRTP(layer int) {
 			return
 		}
 
-		w.locks[layer].Lock()
+		if w.isSimulcast {
+			pts := w.pendingTracks[layer].Load().([]*DownTrack)
+			if len(pts) > 0 {
+				if pkt.KeyFrame {
+					w.Lock()
+					for _, dt := range pts {
+						w.DeleteDownTrack(dt.CurrentSpatialLayer(), dt.peerID)
+						w.storeDownTrack(layer, dt)
+						dt.SwitchSpatialLayerDone(int32(layer))
 
-		if w.isSimulcast && len(w.pendingTracks[layer]) > 0 {
-			if pkt.KeyFrame {
-				for _, dt := range w.pendingTracks[layer] {
-					w.downTracks[layer] = append(w.downTracks[layer], dt)
-					w.DeleteDownTrack(dt.CurrentSpatialLayer(), dt.peerID)
-					dt.SwitchSpatialLayerDone()
+					}
+					w.pendingTracks[layer].Store(make([]*DownTrack, 0, 10))
+					w.Unlock()
+				} else {
+					w.SendRTCP(pli)
 				}
-				w.pendingTracks[layer] = w.pendingTracks[layer][:0]
-			} else {
-				w.SendRTCP(pli)
 			}
 		}
 
-		for _, dt := range w.downTracks[layer] {
-			if err = dt.WriteRTP(pkt); err != nil {
+		dts := w.downTracks[layer].Load().([]*DownTrack)
+		for _, dt := range dts {
+			if err = dt.WriteRTP(pkt, layer); err != nil {
+				if err == io.EOF && err == io.ErrClosedPipe {
+					w.Lock()
+					w.DeleteDownTrack(layer, dt.id)
+					w.Unlock()
+				}
 				log.Error().Err(err).Str("id", dt.id).Msg("Error writing to down track")
 			}
 		}
-		w.locks[layer].Unlock()
 	}
 
 }
 
 // closeTracks close all tracks from Receiver
 func (w *WebRTCReceiver) closeTracks() {
-	for idx, layer := range w.downTracks {
-		w.locks[idx].Lock()
-		for _, dt := range layer {
+	for idx, a := range w.available {
+		if !a.get() {
+			continue
+		}
+		for _, dt := range w.downTracks[idx].Load().([]*DownTrack) {
 			dt.Close()
 		}
-		w.downTracks[idx] = w.downTracks[idx][:0]
-		w.locks[idx].Unlock()
 	}
 	w.nackWorker.StopWait()
 	if w.onCloseHandler != nil {
 		w.onCloseHandler()
 	}
 }
 
-func downTrackSubscribed(dts []*DownTrack, dt *DownTrack) bool {
+func (w *WebRTCReceiver) downTrackSubscribed(layer int, dt *DownTrack) bool {
+	dts := w.downTracks[layer].Load().([]*DownTrack)
 	for _, cdt := range dts {
 		if cdt == dt {
 			return true
 		}
 	}
 	return false
 }
+
+func (w *WebRTCReceiver) storeDownTrack(layer int, dt *DownTrack) {
+	dts := w.downTracks[layer].Load().([]*DownTrack)
+	ndts := make([]*DownTrack, len(dts)+1)
+	copy(ndts, dts)
+	ndts[len(ndts)-1] = dt
+	w.downTracks[layer].Store(ndts)
+}
+
+func (w *WebRTCReceiver) storePendingTrack(layer int, dt *DownTrack) {
+	w.Lock()
```

Comment:
what's the typical convention that's used here? for where the locks should be? just wondering why `storeDownTrack` expects callers to lock, while `storePendingTrack` locks within the function.

Reply:
Since pending track are stored only in one function I put the lock there, but store to downtrack are called in other places, so to not lock many times for example:
```go
					for _, dt := range pts {
						w.DeleteDownTrack(dt.CurrentSpatialLayer(), dt.peerID)
						w.storeDownTrack(layer, dt)
						dt.SwitchSpatialLayerDone(int32(layer))

					}
```
Above code would require to get the locks many times.

Reply:
that makes sense. I think functionality-wise that's optimal. I do wonder if there's a way to make this more obvious to the reader. I've seen others use the convention of public/private functions where locking is handled in the public.

Reply:
Yeah I was thinking on doing that for `DeleteDownTrack` other publics methods are correctly protected